# Broadcasting example

### Calories from Carb, Proteins, Fats in 100g of diffrent food:

 | Apples | Beef | Eggs | Potatoes
 ---: | ---: | ---: | ---: | ---:
**Carb** | 56.0 | 0.0 | 4.4 | 68.0
**Protein** | 1.2 | 104.0 | 52.0 | 8.0
**Fat** | 1.8 | 135.0 | 99.0 | 0.9
**Total Cal** | **59.0** | **239.0** | **155.4** | **76.9**

Using this data we need to calculate the percentage of calories from carb, protein and fat whitout a for-loop





In [20]:
import numpy as np

A = np.array([[56.0, 0.0, 4.4, 68.0],
              [1.2, 104.0, 52.0, 8.0],
              [1.8, 135.0, 99.0, 0.9]])

print(A)

[[  56.     0.     4.4   68. ]
 [   1.2  104.    52.     8. ]
 [   1.8  135.    99.     0.9]]


In [21]:
cal = A.sum(axis=0) # cal = np.sum(A, axis=0), axis=0 to sum verticaly
print(cal)

[  59.   239.   155.4   76.9]


In [22]:
percentage = 100 * A/cal.reshape(1, 4) #[[  59.   239.   155.4   76.9]]
print(percentage)

[[ 94.91525424   0.           2.83140283  88.42652796]
 [  2.03389831  43.51464435  33.46203346  10.40312094]
 [  3.05084746  56.48535565  63.70656371   1.17035111]]


### Matrices

$
\begin{bmatrix}
    1 \\
    2 \\
    3 \\
    4 \\
\end{bmatrix} + 100
 = 
\begin{bmatrix}
    1 \\
    2 \\
    3 \\
    4 \\
\end{bmatrix} + 
\begin{bmatrix}
    100 \\
    100 \\
    100 \\
    100 \\
\end{bmatrix} =
\begin{bmatrix}
    101 \\
    102 \\
    103 \\
    104 \\
\end{bmatrix}
$

$
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6 \\
\end{bmatrix} + 
\begin{bmatrix}
    100 & 200 & 300 \\
\end{bmatrix} = 
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6 \\
\end{bmatrix} + 
\begin{bmatrix}
    100 & 200 & 300 \\
    100 & 200 & 300 \\
\end{bmatrix} = 
\begin{bmatrix}
    101 & 202 & 303 \\
    104 & 205 & 306 \\
\end{bmatrix} 
$

$
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6 \\
\end{bmatrix} + 
\begin{bmatrix}
    100 \\
    200 \\
\end{bmatrix} = 
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6 \\
\end{bmatrix} + 
\begin{bmatrix}
    100 & 100 & 100\\
    200 & 200 & 200\\
\end{bmatrix} = 
\begin{bmatrix}
    101 & 102 & 103 \\
    204 & 205 & 206 \\
\end{bmatrix} 
$


# Tips and Tricks

Do not use data structure without row and column such as `(n,)` use `(n, 1)` or `(1, n)` instead


```python
a = np.random.rand(5)
# a.shape = (5,) 
# a rank 1 array
# DO NOT use this range 1 array

a = np.random.rand(5, 1)
# a.shape = (5, 1) 
# a column vector

a = np.random.rand(1, 5)
# a.shape = (1, 5) 
# a row vector

# use assert to check the matrix dimension (very cheap to compute)
assert(a.shape == (5,1))

# use reshape to convert rank 1 array to vector to avoid errors (very cheap to comput)
a.reshape((5,1))

```

In [44]:
import numpy as np

a = np.random.randn(5)

In [45]:
print(a)

[-0.71289407 -0.23760186  0.51380417 -0.34805323  0.31586985]


In [46]:
print(a.shape)

(5,)


In [47]:
print(a.T) # exactly the same thing with a

[-0.71289407 -0.23760186  0.51380417 -0.34805323  0.31586985]


In [48]:
print(np.dot(a, a.T))

1.04958214437


In [49]:
a = np.random.randn(5, 1)
print(a)

[[-2.14753341]
 [-1.74908005]
 [-0.4774603 ]
 [ 0.14832567]
 [-0.48723265]]


In [50]:
print(a.T)

[[-2.14753341 -1.74908005 -0.4774603   0.14832567 -0.48723265]]


In [51]:
print(np.dot(a, a.T))

[[ 4.61189975  3.75620784  1.02536194 -0.31853432  1.0463484 ]
 [ 3.75620784  3.05928101  0.83511628 -0.25943346  0.85220891]
 [ 1.02536194  0.83511628  0.22796834 -0.07081962  0.23263425]
 [-0.31853432 -0.25943346 -0.07081962  0.0220005  -0.07226911]
 [ 1.0463484   0.85220891  0.23263425 -0.07226911  0.23739566]]
